text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+
+*We want to implement this RFC to prevent (or at least greatly reduce) the occurrences of such issues.*
+
+## (2) Refactor CX Protocol versioning
+
+We've talked about the mechanism we currently use to perform protocol version changes quite a bit. The process and the expected output aren't very clear, and we seem to be always trying to understand it again and again.
+This might warrant a re-think/re-design, to make it much more ergonomic and developer friendly. 
+
+However, performing changes on such a crucial part of the system, shouldn't be done without a proper validation mechanism.
+
+*We want to implement this RFC so that we are able to iterate and change the versioning mechanism itself* 
+
+Actually, We have already acknowledged the need for this a while back:
+
+- [aws/aws-cdk: cx-protocol backwards compatibility integration tests](https://github.com/aws/cdk-ops/issues/365)
+- [aws/aws-cdk: Design: Strategy for automated backwards-compatibility testing](https://github.com/aws/cdk-ops/issues/233)
+
+# Design Summary
+
+In order to ensure compatibility between the CLI and the Framework, we have to actually run the CLI against an old framework version, and vice versa.
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implemented in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using an **old** CLI version and a **new** Framework version.*
+- *Mark a Framework version as incompatible with the latest published CLI version.*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
+
+# Detailed Design
+
+As mentioned, everything is based on our ability to run integration tests with various version permutations. We introduce a new function, `run_integration_tests_perm` which magically does what we need, as well as some useful constants.
+
+The design is best explained by simply looking at the proposed code changes. Most of what you need to know is in the inline comments of the code.
+
+**[common.bash](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/integ/cli/common.bash)**
```

Comment:
Adam has been writing new integ tests using Jest (instead of bash) since what we're building is starting to outstrip the complexity threshold that bash scripts should have.

Maybe it's worth discussing rewriting the test suite in a proper programming language with a proper (and nice!) test runner, like TS/JS + Jest.

Reply:
I like this idea, definitely worth exploring.